aten/src/ATen/cuda/detail/CUDAHooks.cpp

```diff
@@ -459,6 +462,134 @@ void CUDAHooks::deviceSynchronize(DeviceIndex device_index) const {
   c10::cuda::device_synchronize();
 }
 
+std::tuple<size_t, size_t, ptrdiff_t, std::string, std::string, std::string, uint64_t, bool>
+CUDAHooks::StorageShareDevice(const c10::Storage& storage) const {
+    // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+    auto shandle = c10::cuda::CUDACachingAllocator::shareIpcHandle(storage.mutable_data());
+    size_t ipc_memory_handle_size = shandle.handle.size();
+    size_t ipc_event_handle_size = sizeof(cudaIpcEventHandle_t);
+    std::string memory_handle(ipc_memory_handle_size, '\0');
+    std::string event_handle(ipc_event_handle_size, '\0');
+
+    uint64_t ref_counter_offset;
+    bool event_sync_required;
+    ptrdiff_t offset_bytes = shandle.offset;
+    std::memcpy(memory_handle.data(), shandle.handle.c_str(), shandle.handle.size());
+
+    // Put Storage Data behind new ref counting context
+    // See Note [CUDA IPC Refcounting implementation explained]
+    at::DataPtr sent_data_ptr = torch::GetNewRefCountedSentData(
+        storage.mutable_data(), storage.device());
+    auto old_data_ptr = storage.set_data_ptr(std::move(sent_data_ptr));
+    auto sent_data =
+        static_cast<torch::CudaIPCSentData*>(storage.data_ptr().get_context());
+    sent_data->set_original_ptr(std::move(old_data_ptr));
+    std::string ref_counter(sizeof(sent_data->handle()), '\0');
+    std::memcpy(ref_counter.data(), (sent_data->handle()).c_str(), sizeof(sent_data->handle()));
```

Comment:
This looks wrong. You're copying sizeof(sent_data->handle()) which is the size of the string object itself, not the underlying data.